src/compiler/checker.ts

```diff
@@ -6084,9 +6084,9 @@ namespace ts {
         }
 
         function isContextSensitiveFunctionLikeDeclaration(node: FunctionLikeDeclaration) {
-            const areAllParametersUntyped = !forEach(node.parameters, p => p.type);
+            const hasUntypedParameterOrNoParameter = node.parameters.length === 0 || forEach(node.parameters, p => !p.type);
             const isNullaryArrow = node.kind === SyntaxKind.ArrowFunction && !node.parameters.length;
-            return !node.typeParameters && areAllParametersUntyped && !isNullaryArrow;
+            return !node.typeParameters && hasUntypedParameterOrNoParameter && !isNullaryArrow;

```

Comment:
An issue here is that a function expression with fully annotated parameters is not considered context sensitive, which means it won't have a type inferred for `this`.